tests/testthat/test-proxy-restore.R

```diff
@@ -129,3 +129,43 @@ test_that("names<- is not called with partial data (#1108)", {
   vec_c(x, x)
   expect_equal(values, list(c("a", "b", "a", "b")))
 })
+
+test_that("proxy and restore pass `recurse` through `...`", {
+  new_recursive_rcrd <- function(x) {
+    new_rcrd(
+      list(field = x),
+      class = "my_recursive_rcrd"
+    )
+  }
+
+  internal <- new_rcrd(list(internal_field = 1:2))
+  x <- new_recursive_rcrd(data_frame(col = internal))
+
+  local_methods(
+    vec_proxy.my_recursive_rcrd = function(x, ...) {
+      vec_proxy(field(x, "field"), ...)
+    },
+    vec_restore.my_recursive_rcrd = function(x, to, ...) {
+      out <- vec_restore(x, field(to, "field"), ...)
+      new_recursive_rcrd(out)
+    }
+  )
+
+  proxy <- vec_proxy(x, recurse = TRUE)
+  exp <- data_frame(col = data_frame(internal_field = 1:2))
+  expect_equal(proxy, exp)
```

Comment:
It might be worth testing that the `recurse = FALSE` case _doesn't_ proxy recursively, just to ensure the behavior we expect to happen is actually happening